src/daft-core/src/datatypes/matching.rs

```diff
@@ -1,293 +1,362 @@
-        // an exhaustive match across **all** Daft types.
-        // _ => panic!("{:?} not implemented for with_match_daft_types", $key_type)
-    }
-})}
+macro_rules! with_match_daft_types {
+    (
+        $key_type:expr
+        , |$_:tt $T:ident| $($body:tt)*
+        $(,)?
+    ) => {{
+        macro_rules! __with_ty__ {( $_ $T:ident ) => ( $($body)* )}
+        use $crate::datatypes::*;
+        use $crate::datatypes::DataType;
+
+        match $key_type {
+            // Float16 => unimplemented!("Array for Float16 DataType not implemented"),
+            DataType::Binary => __with_ty__! { BinaryType },
+            DataType::Boolean => __with_ty__! { BooleanType },
+            DataType::Date => __with_ty__! { DateType },
+            DataType::Decimal128(..) => __with_ty__! { Decimal128Type },
+            DataType::Duration(_) => __with_ty__! { DurationType },
+            DataType::Embedding(..) => __with_ty__! { EmbeddingType },
+            DataType::Extension(_, _, _) => __with_ty__! { ExtensionType },
+            DataType::FixedShapeImage(..) => __with_ty__! { FixedShapeImageType },
+            DataType::FixedShapeSparseTensor(..) => __with_ty__! { FixedShapeSparseTensorType },
+            DataType::FixedShapeTensor(..) => __with_ty__! { FixedShapeTensorType },
+            DataType::FixedSizeBinary(_) => __with_ty__! { FixedSizeBinaryType },
+            DataType::FixedSizeList(_, _) => __with_ty__! { FixedSizeListType },
+            DataType::Float32 => __with_ty__! { Float32Type },
+            DataType::Float64 => __with_ty__! { Float64Type },
+            DataType::Image(..) => __with_ty__! { ImageType },
+            DataType::Int16 => __with_ty__! { Int16Type },
+            DataType::Int32 => __with_ty__! { Int32Type },
+            DataType::Int64 => __with_ty__! { Int64Type },
+            DataType::Int8 => __with_ty__! { Int8Type },
+            DataType::Interval => __with_ty__! { IntervalType },
+            DataType::List(_) => __with_ty__! { ListType },
+            DataType::Map{..} => __with_ty__! { MapType },
+            DataType::Null => __with_ty__! { NullType },
+            DataType::SparseTensor(..) => __with_ty__! { SparseTensorType },
+            DataType::Struct(_) => __with_ty__! { StructType },
+            DataType::Tensor(..) => __with_ty__! { TensorType },
+            DataType::Time(_) => __with_ty__! { TimeType },
+            DataType::Timestamp(_, _) => __with_ty__! { TimestampType },
+            DataType::UInt16 => __with_ty__! { UInt16Type },
+            DataType::UInt32 => __with_ty__! { UInt32Type },
+            DataType::UInt64 => __with_ty__! { UInt64Type },
+            DataType::UInt8 => __with_ty__! { UInt8Type },
+            DataType::Unknown => unimplemented!("array for unknown datatype not implemented"),
+            DataType::Utf8 => __with_ty__! { Utf8Type },
+            #[cfg(feature = "python")]
+            DataType::Python => __with_ty__! { PythonType },
```

Comment:
if you want to be even more idiomatic can do $crate::DataType::{x}

similar with these (assuming same crate)

I'd recommend this

Reply:
I'm not sure I'm following. Could you show me an example, please?

Reply:
actually this does not matter. I misread; we are importing with

`use $crate::datatypes::DataType;`